cookbook/core/type_system/pytorch_types.py

```diff
@@ -0,0 +1,192 @@
+
+# %%
+# Checkpoint
+# ----------
+#
+# ``PyTorchCheckpoint`` is a special type of checkpoint that is used to serialize and deserialize PyTorch models.
+# It checkpoints ``torch.nn.Module``'s state, hyperparameters, and optimizer state.
+# The module checkpoint differs from the standard checkpoint in that it checkpoints the module's ``state_dict``.
+# Hence, when restoring the module, the module's ``state_dict`` needs to be used in conjunction with the actual module.
+# As per PyTorch `docs <https://pytorch.org/tutorials/beginner/saving_loading_models.html#save-load-entire-model>`__, it is recommended to
+# store the module's ``state_dict`` rather than the module itself. However, the serialization should work either way.
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.optim as optim
+from dataclasses import dataclass
+
+from dataclasses_json import dataclass_json
+from flytekit.extras.pytorch import PyTorchCheckpoint
+
+
+@dataclass_json
+@dataclass
+class Hyperparameters:
+    epochs: int
+    loss: float
+
+
+class Net(nn.Module):
+    def __init__(self):
+        super(Net, self).__init__()
+        self.conv1 = nn.Conv2d(3, 6, 5)
+        self.pool = nn.MaxPool2d(2, 2)
+        self.conv2 = nn.Conv2d(6, 16, 5)
+        self.fc1 = nn.Linear(16 * 5 * 5, 120)
+        self.fc2 = nn.Linear(120, 84)
+        self.fc3 = nn.Linear(84, 10)
+
+    def forward(self, x):
+        x = self.pool(F.relu(self.conv1(x)))
+        x = self.pool(F.relu(self.conv2(x)))
+        x = x.view(-1, 16 * 5 * 5)
+        x = F.relu(self.fc1(x))
+        x = F.relu(self.fc2(x))
+        x = self.fc3(x)
+        return x
+
+
+@task
+def generate_model(hyperparameters: Hyperparameters) -> PyTorchCheckpoint:
+    bn = Net()
+    optimizer = optim.SGD(bn.parameters(), lr=0.001, momentum=0.9)
+    return PyTorchCheckpoint(
+        module=bn, hyperparameters=hyperparameters, optimizer=optimizer
+    )
+
+
+@task
+def load(checkpoint: PyTorchCheckpoint):
+    new_bn = Net()
+    new_bn.load_state_dict(checkpoint["module_state_dict"])
+    optimizer = optim.SGD(new_bn.parameters(), lr=0.001, momentum=0.9)
+    optimizer.load_state_dict(checkpoint["optimizer_state_dict"])
+
+
+@workflow
+def pytorch_checkpoint_wf():
+    checkpoint = generate_model(hyperparameters=Hyperparameters(epochs=10, loss=0.1))
+    load(checkpoint=checkpoint)
+
+
+# %%
+# .. note::
+#   ``PyTorchCheckpoint`` supports serializing hyperparameters of type ``dict``, ``NamedTuple``, and ``dataclass``.
+#
+# Auto GPU to CPU & CPU to GPU Conversion
+# ---------------------------------------
+#
+# Not all PyTorch computations need to run on a GPU. There are some cases where it is beneficial to move the computation to the CPU after say,
+# the model is trained on a GPU. To avail the GPU power when the computations are run on a GPU, we do ``to(torch.device("cuda"))``.
+# To use the GPU-variables on a CPU, we need to move the variables to a CPU using ``to(torch.device("cpu"))`` construct.
```

Comment:
To avail the GPU power when required, use ``to(torch.device("cuda"))``.
To use the GPU-variables on a CPU, move the variables to a CPU using ``to(torch.device("cpu"))`` construct.

Reply:
The tone here is more of instructing the user as to how PyTorch does this. So I guess "we" is fine in this case.